core/metacling/src/TClingGenericInfo.h

```diff
@@ -0,0 +1,46 @@
+/// \file TClingGenericInfo.h
+///
+/// \brief The file contains a base class of TCling*Info classes.
+///
+/// \author Vassil Vassilev <vvasilev@cern.ch>
+///
+/// \date March, 2019
+///
+/*************************************************************************
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_TClingGenericInfo
+#define ROOT_TClingGenericInfo
+
+#include <string>
+
+namespace clang {
+   class Decl;
+}
+
+class TClingGenericInfo {
+protected:
+   const clang::Decl* fDecl = nullptr;
+   std::string fNameCache;
+public:
+   TClingGenericInfo(const clang::Decl* D) : fDecl(D) {}
+   TClingGenericInfo(const TClingGenericInfo& other) {
+      fDecl = other.fDecl;
+      fNameCache = ""; // we must not copy the cache.
+   }
+   virtual ~TClingGenericInfo();
+
+   virtual const clang::Decl* GetDecl() const { return fDecl; }
```

Comment:
why is the const version virtual but the non-const version is not virtual?

Reply:
The non-const is essentially virtual, as it forwards to the virtual one. This way I want to reuse implementation and disallow people to override the non-cost version.